### ISSUE TITLE
FreeBSD: apps.plugin reports spikes and apps.cpu less user CPU

### ISSUE THREAD

**Issue by Dennis4b** [netdata/netdata#3245]
Thank you for supporting FreeBSD.
I'm having some issues where `apps.cpu_system` shows huge overflows; instead of something like "500000" (meaning 50ms at `RATES_DETAIL`) you sporadically get values like "18432476416962" which blow out the chart.

I've found the following error in timing calculations, it doesn't completely solve my issue yet but it's probably a start.

```
diff --git a/src/apps_plugin.c b/src/apps_plugin.c
index 9e122402..52aaed32 100644
--- a/src/apps_plugin.c
+++ b/src/apps_plugin.c
@@ -955,7 +955,7 @@ static inline int read_proc_pid_stat(struct pid_stat *p, void *ptr) {
     pid_incremental_rate(stat, p->utime,   (kernel_uint_t)proc_info->ki_rusage.ru_utime.tv_sec * 100 + proc_info->ki_rusage.ru_utime.tv_usec / 10000);
     pid_incremental_rate(stat, p->stime,   (kernel_uint_t)proc_info->ki_rusage.ru_stime.tv_sec * 100 + proc_info->ki_rusage.ru_stime.tv_usec / 10000);
     pid_incremental_rate(stat, p->cutime,  (kernel_uint_t)proc_info->ki_rusage_ch.ru_utime.tv_sec * 100 + proc_info->ki_rusage_ch.ru_utime.tv_usec / 10000);
-    pid_incremental_rate(stat, p->cstime,  (kernel_uint_t)proc_info->ki_rusage_ch.ru_stime.tv_sec * 100 + proc_info->ki_rusage_ch.ru_utime.tv_usec / 10000);
+    pid_incremental_rate(stat, p->cstime,  (kernel_uint_t)proc_info->ki_rusage_ch.ru_stime.tv_sec * 100 + proc_info->ki_rusage_ch.ru_stime.tv_usec / 10000);
 
     p->num_threads      = proc_info->ki_numthreads;
 

```

**Comment by ktsaou** [comment]
thanks!

@vlvkobal please have a look.

**Comment by ktsaou** [comment]
in pr #3276 

Thank you for reporting it.

**Comment by ktsaou** [comment]
There is another report about apps.plugins: #2646 

I found this freebsd code: https://github.com/firehol/netdata/blob/cc664a373e116c1050c5818cdbe3f06a148becad/src/apps_plugin.c#L2133-L2136

The above says: loop through pid `1` to `procnum - INIT_PID - 1`. So, this loop looses 2 processes (`INIT_PID = 1`): the first and the last.

If FreeBSD provides the list sorted, pid 0 should be harmless and ignorable. However the last pid that is lost may have affected the accumulations of exited children into their parents and could result in spikes.

I made a few more changes to improve the code in case it was missing pids under freebsd due to the way sysctl calls are made.

I will merge it in a while.

**Comment by ktsaou** [comment]
merged it.

**Comment by Dennis4b** [comment]
Quick followup:

First of all thank you!

Graph looks a lot better now, there are still occasional peaks that are not supported by the general CPU utilization graph, more in the range of 600-1000% (on a 4 core system where 400% should be max), so there's still something not _quite_ right, but overall the graphs are very useful and it's quite brilliant to be able to visualize resource usage by your own defined process groups.

**Comment by ktsaou** [comment]
> there are still occasional peaks that are not supported by the general CPU utilization graph, more in the range of 600-1000% (on a 4 core system where 400% should be max), so there's still something not quite right

Is it on random targets or it happens on specific ones?

I think this can only happen by the accumulation of children process resources into their parents. You can disable this feature with the option `without-childs` in the apps.plugin section of `netdata.conf`:

```
[plugin:apps]
	# update every = 1
	command options = without-childs
```

There is a compile time decision about FreeBSD, that all the process utilization is given to apps.plugin (by the kernel) instantly. If we disable this, apps.plugin has code to normalize the exited children utilization, so that it will never exceed the total CPU utilization of the system. This was implemented for Linux, that apps.plugin needs some time to walk through the process tree, so the exceeded children utilization is frequenty exceeding the total CPU utilization of the system.

But before going so drastic, I suggest to spend some time understanding the issue. Can you help?

**Comment by Dennis4b** [comment]
The peaks do occur with those apps that spawn child processes, I never see them with for example the  postgresql backends.

When they do happen then either one of, or both of, the user and system cpu may be at well above 400%, so I agree it probably has something to do with child process accumulation and not a simple errant calculation.

If my `web` app for example spawns lots of `pdftk` processes to manipulate PDFs, and I have an application group for both (so one for `web` _and_ one for `pdftk`), what should actually happen when `without-childs` is not set? Is the `pdftk` time then counted for both app groups or is the `pdtk` not considered as a separate process because it is a child?

Is `without-childs` then the more logical choice if I am interested in the spawned (child) processes separately?

I'm very happy to help and I'll be able to collect more data during next week with peak usage, right now nothing interesting is happening because of the weekend. I'll observe for a bit.

**Comment by ktsaou** [comment]
`without-childs` will have the following effects:

1. childs that have spawned and exited within the same iteration, will not provide any input to their parent (they will be invisible)

2. childs that had been running for at least one iteration but found exited that the time apps.plugin checked, will not provide any input to their parent (their tail utilization will be invisible).

So, apps.plugin will work mostly like `top`: only processes found running will be accumulated - they processes that started and exited between the data collection points will no influence the output of the plugin.

To understand the problem, take this example:

1. At `t + 0` process `A` is started

2. At `t + 10` process `A` starts process `B`

3. For 5 iterations (up to `t + 15`) both are running and each process is found by apps.plugin and they are both aggregated to the same metric.

4. At `t + 15` process `B` exits. This is the tricky part. The exited children counters of process `A` take **instantly the whole utilization of process `B` (for its entire lifetime)**. Of course this would lead to an unrealistic spike. So, apps.plugin subtracts from the cpu utilization of the exited children of process `A`, all the utilization it has collected so far for process `B`.

Let me build a small test case for that, to verify it works on freebsd...

**Comment by ktsaou** [comment]
Well, I create a simple bash test:

<details>
<summary>click to see test-apps.plugin-exited-children.sh</summary>

```sh
#!/usr/bin/env bash

busy_loop() {
	local count=${1} c=0 ppid=${2} pppid=${3}

	while [ ${count} -gt 0 ]
	do
		count=$((count - 1))
		c=$((c + 1))
		echo "${count} >> ${c}" >/dev/null
	done

	echo "loops ${c}, pid ${BASHPID}, ppid ${ppid}, pppid ${pppid}"
}

do_one_set() {
	local x loop=0 s=${1} pid=${BASHPID} ppid=${2}

	echo >&2
	echo >&2 "Started set ${s}, on pid ${pid}, ppid ${ppid}"

	for x in 1000 10000 25000 50000 62500 75000 87500 100000 200000 300000 400000 500000
	do
		loop=$((loop + 1))
		printf >&2 "busy loop Set No ${s}, Loop No ${loop} for $x iterations... "
		o="$(busy_loop ${x} ${pid} ${ppid})"
		printf >&2 "OK (${o})\n"
		sleep 5
	done
}

s=0
p=${BASHPID}
while [ 1 ]
do
	s=$((s + 1))
	z="$(do_one_set ${s} ${p})"
	sleep 5
done

```

In `/etc/netdata/apps_groups.conf` I added before anything else:

```
test: *test-apps.plugin-exited-children.sh*
```

and this is some sample output:

```
Started set 1, on pid 1990, ppid 1989
busy loop Set No 1, Loop No 1 for 1000 iterations... OK (loops 1000, pid 1991, ppid 1990, pppid 1989)
busy loop Set No 1, Loop No 2 for 10000 iterations... OK (loops 10000, pid 1993, ppid 1990, pppid 1989)
busy loop Set No 1, Loop No 3 for 25000 iterations... OK (loops 25000, pid 1995, ppid 1990, pppid 1989)
busy loop Set No 1, Loop No 4 for 50000 iterations... OK (loops 50000, pid 1997, ppid 1990, pppid 1989)
busy loop Set No 1, Loop No 5 for 62500 iterations... OK (loops 62500, pid 1999, ppid 1990, pppid 1989)
busy loop Set No 1, Loop No 6 for 75000 iterations... OK (loops 75000, pid 2001, ppid 1990, pppid 1989)
busy loop Set No 1, Loop No 7 for 87500 iterations... OK (loops 87500, pid 2005, ppid 1990, pppid 1989)
busy loop Set No 1, Loop No 8 for 100000 iterations... OK (loops 100000, pid 2007, ppid 1990, pppid 1989)
busy loop Set No 1, Loop No 9 for 200000 iterations... OK (loops 200000, pid 2009, ppid 1990, pppid 1989)
busy loop Set No 1, Loop No 10 for 300000 iterations... OK (loops 300000, pid 2011, ppid 1990, pppid 1989)
busy loop Set No 1, Loop No 11 for 400000 iterations... OK (loops 400000, pid 2013, ppid 1990, pppid 1989)
busy loop Set No 1, Loop No 12 for 500000 iterations... OK (loops 500000, pid 2015, ppid 1990, pppid 1989)

Started set 2, on pid 2018, ppid 1989
busy loop Set No 2, Loop No 1 for 1000 iterations... OK (loops 1000, pid 2019, ppid 2018, pppid 1989)
busy loop Set No 2, Loop No 2 for 10000 iterations... OK (loops 10000, pid 2021, ppid 2018, pppid 1989)
busy loop Set No 2, Loop No 3 for 25000 iterations... OK (loops 25000, pid 2023, ppid 2018, pppid 1989)
busy loop Set No 2, Loop No 4 for 50000 iterations... OK (loops 50000, pid 2025, ppid 2018, pppid 1989)
busy loop Set No 2, Loop No 5 for 62500 iterations... OK (loops 62500, pid 2027, ppid 2018, pppid 1989)
busy loop Set No 2, Loop No 6 for 75000 iterations... OK (loops 75000, pid 2029, ppid 2018, pppid 1989)
busy loop Set No 2, Loop No 7 for 87500 iterations... OK (loops 87500, pid 2031, ppid 2018, pppid 1989)
busy loop Set No 2, Loop No 8 for 100000 iterations... OK (loops 100000, pid 2033, ppid 2018, pppid 1989)
busy loop Set No 2, Loop No 9 for 200000 iterations... OK (loops 200000, pid 2035, ppid 2018, pppid 1989)
busy loop Set No 2, Loop No 10 for 300000 iterations... OK (loops 300000, pid 2037, ppid 2018, pppid 1989)
busy loop Set No 2, Loop No 11 for 400000 iterations... OK (loops 400000, pid 2039, ppid 2018, pppid 1989)
busy loop Set No 2, Loop No 12 for 500000 iterations... OK (loops 500000, pid 2052, ppid 2018, pppid 1989)
```

</details>

---

Here is it shows on a single CPU freebsd VM:

### freebsd `system.cpu`

![screenshot from 2018-01-20 14-45-42](https://user-images.githubusercontent.com/2662304/35183557-daf7dcdc-fdf0-11e7-9f10-259ac674f3aa.png)

### freebsd `apps.cpu`

![screenshot from 2018-01-20 14-46-05](https://user-images.githubusercontent.com/2662304/35183561-ef9208a2-fdf0-11e7-9727-7b96d78a095c.png)

2 problems:

1. it spikes above 100%
2. the legit 100% is reported as 80%

---

The same on linux (same setup, 1 CPU core)

### linux `system.cpu`

![screenshot from 2018-01-20 14-47-03](https://user-images.githubusercontent.com/2662304/35183572-1797e2fe-fdf1-11e7-981b-63b267a12d8a.png)

### linux `apps.cpu`

![screenshot from 2018-01-20 14-47-20](https://user-images.githubusercontent.com/2662304/35183573-1c6adfe8-fdf1-11e7-9cbd-a13e4db90802.png)

Here the normalization absorbed a few spikes entirely (clearly visible at the last set), but apps.plugin never reported CPU utilization above the system total.

**Comment by ktsaou** [comment]
In more detail for freebsd:

## problem 1: per process `user` cpu is less by 25%

![screenshot from 2018-01-20 15-17-36](https://user-images.githubusercontent.com/2662304/35183837-13c89fa2-fdf5-11e7-9b34-85a1d54fc31f.png)

it seems that the reported per process `user` process **is constantly less by 25%**.
Here is again the system one:

![screenshot from 2018-01-20 14-45-42](https://user-images.githubusercontent.com/2662304/35183557-daf7dcdc-fdf0-11e7-9f10-259ac674f3aa.png)

So, per pid `user` CPU should peak at around 80% but it peaks at around 60%.

## problem 2: exited children cpu (`user` and `system`) spike

It seems the problem exists only when a process starts and exits within the same `apps.plugin` iteration.
In this case, `apps.plugin` does not know anything about the pid that started and exited. It trusts the kernel to report the right value.

@vlvkobal I really need your help here.

**Comment by vlvkobal** [comment]
Got the same problem on Linux VM with 1 CPU:
![image](https://user-images.githubusercontent.com/22008543/47991354-99181580-e0f2-11e8-9b79-969e37c8b478.png)

**Comment by vlvkobal** [comment]
@Dennis4b The issue should be fixed in #5038. Please test it.
